File: ruby/abc222_e.rb
n,m,k = gets.split(" ").map(&:to_i)
a = gets.split(" ").map{|e|e.to_i - 1}
$l = Array.new(n+1){[]}
(n-1).times do |i|
	x,y = gets.split(" ").map(&:to_i)
	x-=1;y-=1
	$l[x] << [y,i]
	$l[y] << [x,i]
end

$cnt = Array.new(n-1,0)
def dfs(from,to,pre)
	return true if from == to
	for ver,ind in $l[from]
		next if ver==pre
		if dfs(ver,to,from)
			$cnt[ind] += 1
			return true
		end
	end
	return false
end

for i in 1...m
	dfs(a[i-1],a[i],-1)
end

total = $cnt.sum
if (total+k)&1==1 || (total+k)<0
	p 0
	exit
end

MOD  = 998244353
dp = [0]*100001
dp[0] = 1
(n-1).times do |i|
	100000.downto($cnt[i]) do |j|
		dp[j] += dp[j-$cnt[i]]
		dp[j-$cnt[i]] %= MOD
	end
end

p dp[(total+k)>>1]

File: python/abc-e/abc222_e.py
n, m, k = map(int, input().split())
a = [*map(int, input().split())]
a = [i - 1 for i in a]
l = [[] for _ in range(n)]
for i in range(n - 1):
    x, y = map(int, input().split())
    x -= 1
    y -= 1
    l[x].append((y, i))
    l[y].append((x, i))

c = [0] * (n - 1)


def dfs(f, t, p):
    if f == t:
        return 1
    for v, i in l[f]:
        if v == p:
            continue
        if dfs(v, t, f):
            c[i] += 1
            return 1
    return 0


for i in range(1, m):
    dfs(a[i - 1], a[i], -1)

s = sum(c)
if (s + k) & 1 or s + k < 0:
    print(0)
    exit()

MOD = 998244353
dp = [0] * 100001
dp[0] = 1
for i in range(n - 1):
    for j in range(100000, c[i] - 1, -1):
        dp[j] += dp[j - c[i]]
        dp[j] %= MOD
        dp[j - c[i]] %= MOD

print(dp[(s + k) >> 1])
